G-API: oneVPL DX11 inference

sivanov-work opened the pull request:
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake


### Build Configuration

```
force_builders=XCustom,Custom Win,Custom Mac
build_gapi_standalone:Linux x64=ade-0.1.1f
build_gapi_standalone:Win64=ade-0.1.1f
build_gapi_standalone:Mac=ade-0.1.1f
build_gapi_standalone:Linux x64 Debug=ade-0.1.1f

Xbuild_image:Custom=centos:7
Xbuildworker:Custom=linux-1
build_gapi_standalone:Custom=ade-0.1.1f

build_image:Custom=ubuntu-openvino-2021.3.0:20.04
Xbuild_image:Custom Win=openvino-2021.2.0
build_image:Custom Mac=openvino-2021.2.0

test_modules:Custom=gapi,python2,python3,java
test_modules:Custom Win=gapi,python2,python3,java
test_modules:Custom Mac=gapi,python2,python3,java

buildworker:Custom=linux-1
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*

build_image:Custom Win=gapi-onevpl-2021.6.0
buildworker:Custom Win=windows-3
build_contrib:Custom Win=OFF
```

sivanov-work commented:
one more commit is expected to fix CustomMac CI in a way to fix comparison between InferenceEngines types

mpashchenkov commented on modules/gapi/src/backends/ie/giebackend.cpp:
```diff
@@ -510,15 +510,51 @@ inline IE::Blob::Ptr extractRemoteBlob(IECallContext& ctx, std::size_t i) {
     using ParamType = std::pair<InferenceEngine::TensorDesc,
```
Not for this changes but:
`auto ie_core = cv::gimpl::ie::wrap::getCore();` - Is unused variable?

sivanov-work replied:
you right, it seems is unused

mpashchenkov commented on modules/gapi/src/backends/ie/giebackend.cpp:
```diff
@@ -510,15 +510,51 @@ inline IE::Blob::Ptr extractRemoteBlob(IECallContext& ctx, std::size_t i) {
+    if (ctx.uu.params.device_id.find("GPU") != std::string::npos) {
+        GAPI_LOG_DEBUG(nullptr, "Extract remote blob for GPU DX11 device_id: " <<
+                                ctx.uu.params.device_id);
+        // despite of layout, blob dimensions always follow in N,C,H,W order
```
Will only 4D blob be supported?

sivanov-work replied:
it depends from openvino  and code taken from  openvino code `make_shared_nv12_blob`

mpashchenkov commented on modules/gapi/src/backends/ie/giebackend.cpp:
```diff
@@ -510,15 +510,51 @@ inline IE::Blob::Ptr extractRemoteBlob(IECallContext& ctx, std::size_t i) {
+        // TODO NV12 surface supported only
+        if (static_cast<InferenceEngine::ColorFormat>(frame_blob_params->second["COLOR_FORMAT"]) ==
+            InferenceEngine::ColorFormat::NV12) {
+            InferenceEngine::Blob::Ptr y_blob =
```
Can be used `InferenceEngine::NV12Blob` for this case?

sivanov-work replied:
it has huge class hierarchy, but actually final blob must be one of descendant of `ClBlob` which represents remote GPU memory. Maybe NV12Blob is present in class inheritance in the middle chain, i'm not sure

mpashchenkov commented on modules/gapi/src/backends/ie/giebackend.cpp:
```diff
@@ -510,15 +510,51 @@ inline IE::Blob::Ptr extractRemoteBlob(IECallContext& ctx, std::size_t i) {
+        if (static_cast<InferenceEngine::ColorFormat>(frame_blob_params->second["COLOR_FORMAT"]) ==
+            InferenceEngine::ColorFormat::NV12) {
+            InferenceEngine::Blob::Ptr y_blob =
+                           std::dynamic_pointer_cast<InferenceEngine::Blob>(
```
Can this be done with some make_..._blob functions? For interest.

sivanov-work replied:
yes it is: 524 - 550 lines related to `make_shared_nv12_blob` API function and to declare that function we need to include specific OV include file with this make* API but this header file has a required dependency from Open**CL** -HPP package which is not a part of open**CV** and should be not according to license type. So all make_* functions from OV simply unwrapped from its definition and copied here

mpashchenkov commented on modules/gapi/src/backends/ie/giebackend.cpp:
```diff
@@ -1114,10 +1163,9 @@ struct InferROI: public cv::detail::KernelTag {
-                                *(ctx->uu.params.input_names.begin()),
-                                IE::make_shared_blob(this_blob, toIE(this_roi)),
-                                *ctx);
+                        setROIBlob(req,
```
I think, the function should be used for InferList and InferList2.
Can GPU ROI support appear in future?

sivanov-work replied:
Do you mean - place such changes SetBlob -> SetROIBlob in other place? could you point me out, please?

mpashchenkov replied:
make_shared_blob for ROIs + SetBlob in InferList and InferList2. Is it necessary align this with SetRoiBlob? Now it is called in one place. 

https://github.com/opencv/opencv/blob/e81100e5e0a5bbad4dc06adb8e443db74acd9c24/modules/gapi/src/backends/ie/giebackend.cpp#L1285

https://github.com/opencv/opencv/blob/e81100e5e0a5bbad4dc06adb8e443db74acd9c24/modules/gapi/src/backends/ie/giebackend.cpp#L1444

mpashchenkov commented on modules/gapi/src/backends/ie/giebackend.cpp:
```diff
@@ -560,6 +596,19 @@ static void setBlob(InferenceEngine::InferRequest& req,
+                       const cv::Rect &roi,
+                       const IECallContext&           ctx) {
+    if (ctx.uu.params.device_id.find("GPU") != std::string::npos) {
+        GAPI_LOG_DEBUG(nullptr, "Skip ROI blob creation for device_id: " <<
```
This can confuse. M b it should be assert.

sivanov-work replied:
definitely no: GPU blob doesn't support ROI but need to continue with SeBlob further
Let's apply limitation that ROI for GPU (right before OpenVINO 2.0 adaptation) must be done on G-API / GSource side manually

OrestChura commented on modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:
```diff
@@ -26,6 +26,31 @@ namespace gapi {
+    } else {
+        if (static_cast<size_t>(std::numeric_limits<mfxU16>::max()) < new_frames_count) {
+            GAPI_LOG_WARNING(nullptr, "Cannot proceed with CfgParam \"" << param_name << "\": " <<
+                                  new_frames_count << ". It must not be equal than " <<
```
`Must not be greater` I believe

sivanov-work replied:
fixed

OrestChura commented on modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.cpp:
```diff
@@ -26,6 +26,31 @@ namespace gapi {
+        GAPI_LOG_DEBUG(nullptr, "mfxFrameAllocRequest overriden by user input: " <<
+                            ", mfxFrameAllocRequest.NumFrameMin: " << request.NumFrameMin <<
+                            ", mfxFrameAllocRequest.NumFrameSuggested: " << request.NumFrameSuggested <<
+                            ", mfxFrameAllocRequest.Type: " << request.Type);
```
Minor: can be aligned

sivanov-work replied:
fixed

sivanov-work commented on modules/gapi/include/opencv2/gapi/streaming/onevpl/cfg_params.hpp:
```diff
@@ -110,6 +110,62 @@ struct GAPI_EXPORTS CfgParam {
     static CfgParam create_implementation(const char* value);
 
+
+    static constexpr const char *vpp_frames_pool_size_name() { return "vpp_frames_pool_size"; }
```
it's ascending to Source itself but not for preprocessing settings: It's parameter similar with `decode_frames_pool_size` old one - and  according to thin optimization  settings

OrestChura replied:
So, it has to be this?  
I thought it is just a name which you use across the code, so you can practically choose anything as a name. I just suggested to keep it similar, is it possible?

sivanov-work replied:
i'm sorry, maybe i was confusing:

> "vpp.FramesPoolSize"

you suggest to name is as "vpp.In.Width", right?
It should not, because params  have different  meaning: "vpp.In.Width" involved in preprocessing, it's configuration param and it used to fill `mfxVideoParam` structure for VPL api; but `vpp_frames_pool_size` determines how much surfaces we  should allocate to make preprocessing fast or minimal memory consumption - it's some kind of optimization params. 
Maybe we should implement different param classes at least: config & optimization. But  before do this, I wish I have get user experience feedback :)

OrestChura replied:
OK!

sivanov-work commented on modules/gapi/samples/onevpl_infer_single_roi.cpp:
```diff
@@ -148,15 +152,25 @@ GAPI_OCV_KERNEL(OCVLocateROI, LocateROI) {
-                           , sqside
-                           };
+        // NB: GPU plugin in InferenceEngine doesn't support ROI at now
+        if (device_id.find("GPU") == std::string::npos) {
```
From my perspective `flag` is redundant here because it duplicates entity device and all information about `flag` can be extracted from `device_id`.  + duplicated entity would become inconsistent when the main entity `device_id` changed for some reason

If you don't like huge check on device type then i propose to introduce global function `bool is_gpu(string)` instead of flag

OrestChura replied:
Yes, these unnesessary repetitions -  I just suggested to reduce their amount.  
The function's call is a repetition too, but it's OK for me - yes, I think it is an OK solution :)

sivanov-work commented on modules/gapi/samples/onevpl_infer_single_roi.cpp:
```diff
@@ -299,6 +324,20 @@ int main(int argc, char *argv[]) {
                                             {"VA_DEVICE", accel_device_ptr} });
```
When this example be extended on other device types you will require more and more binary flags for each device type
`device_id` is more generic and doesn't need any further extension or hacks.

Agree, that this `device_id` may be passed as kernel param. I'll fix it
in other places will introduce function `is_gpu(string)`

OrestChura replied:
OK, got it, then I agree with you

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -261,24 +259,55 @@ mfxStatus VPLDX11AccelerationPolicy::on_alloc(const mfxFrameAllocRequest *reques
     desc.SampleDesc.Count = 1;
     desc.Usage = D3D11_USAGE_DEFAULT;
-    desc.MiscFlags = D3D11_RESOURCE_MISC_SHARED;
+    desc.MiscFlags = 0;// D3D11_RESOURCE_MISC_SHARED;
```
it should have been removed. thanks

sivanov-work replied:
fixed

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/dx11_alloc_resource.cpp:
```diff
@@ -355,13 +356,14 @@ DX11AllocationRecord::~DX11AllocationRecord() {
+
     GAPI_DbgAssert(items != 0 && "Cannot create DX11AllocationRecord with empty items");
     GAPI_DbgAssert(items == staging_textures.size() && "Allocation items count and staging size are not equal");
+    GAPI_DbgAssert(textures.size() != 1 ? items == textures.size() : true && "Allocation items count and staging size are not equal");
```
There are two options: `texture array` and `array of textures`
The first options means that we have **single texture** object and different layers inside it called **subresource_id**.
the other option means: **multiple textures** each with single layer inside it where **subresource_id = 0**
Different options used for different purposes: for decoder part `texture array` is more preferrable for perf, but for VPP part the second option is the only way: some limitations.

So idea of such code is allocate `items` of resources: single texture with `items` layer OR `items` of textures with 1 leayer

OrestChura replied:
OK, great! got the idea, it's cool!

sivanov-work commented on modules/gapi/src/streaming/onevpl/cfg_params_parser.cpp:
```diff
@@ -86,6 +96,76 @@ std::vector<ValueType> get_params_from_string(const std::string& str) {
+        } else if (name == CfgParam::vpp_out_framerate_d_name()) {
+            ret.push_back(creator.create(name,
+                                         static_cast<uint32_t>(strtoul_or_throw(value.c_str())),
+                                         false));
```
It impossible to use switch for non-integral types: `string` is used here

OrestChura replied:
Oh, that's lame.. Sorry

Still, you can unite a lot of ifelse branches which do the exact same push_back

sivanov-work replied:
Sorry, don't catch the idea:
Do you mean

```
if (name == A || name == B || name = C) 
{
push_back(name, **strtoull**);
} else if(name == D) 
{
push_back(name, **strtoll**);
} 
```
is it true?

OrestChura replied:
I think yes, you got me right

The idea is to combine all the conditions with the same action to one condition  
It will increase readability and shorten things

sivanov-work commented on modules/gapi/src/streaming/onevpl/engine/decode/decode_session.hpp:
```diff
@@ -26,9 +26,10 @@ struct IDataProvider;
+class GAPI_EXPORTS LegacyDecodeSession : public EngineSession {
 public:
     friend class VPLLegacyDecodeEngine;
+    friend class VPLLegacyTranscodeEngine; //TODO
```
it should be more public methods in Session instead of private data...
But it looks minor at now 
I shall update comment with TODO

sivanov-work replied:
fixed

OrestChura commented on modules/gapi/src/streaming/onevpl/engine/transcode/transcode_engine_legacy.cpp:
```diff
@@ -0,0 +1,477 @@
+
+ProcessingEngineBase::ExecutionStatus VPLLegacyTranscodeEngine::execute_op(operation_t& op, EngineSession& sess) {
+    return op(sess);
+}
```
Does this method not derived from `VPLLegacyDecodeEngine`? It does there the same thing.  
It's protected there, but is it important to make it private in `VPLLegacyTranscodeEngine`?

EDIT: Why is the method even `virtual` in `ProcessingEngineBase` in the first place? It seems doing the same thing in all `ProcessingEngineBase`,  `VPLLegacyDecodeEngine` and `VPLLegacyTranscodeEngine`

sivanov-work replied:
i don't remember sorry...
may be in unit tests it is used as public.
I'll try to examine it

OrestChura commented on modules/gapi/src/streaming/onevpl/engine/transcode/transcode_engine_legacy.cpp:
```diff
@@ -0,0 +1,477 @@
+}
+
+void VPLLegacyTranscodeEngine::on_frame_ready(LegacyTranscodeSession& sess,
+                                           mfxFrameSurface1* ready_surface)
```
alignment broken a bit

sivanov-work replied:
fixed

sivanov-work commented on modules/gapi/samples/onevpl_infer_single_roi.cpp:
```diff
@@ -45,8 +45,12 @@ const std::string keys =
+    "{ vpp_frames_pool_size         | 0                                         | OneVPL source applies this parameter as preallocated frames pool size for VPP preprocessing results}"
+    "{ source_preproc_enable        | 0                                         | Turn on OneVPL source frame preprocessing using network input description instead of IE plugin preprocessing}";
 
+static uint32_t vpl_source_preproc_enable = 0;
```
you right, it may be on stack, i suppose

sivanov-work commented on modules/gapi/samples/onevpl_infer_single_roi.cpp:
```diff
@@ -45,8 +45,12 @@ const std::string keys =
+    "{ source_preproc_enable        | 0                                         | Turn on OneVPL source frame preprocessing using network input description instead of IE plugin preprocessing}";
 
+static uint32_t vpl_source_preproc_enable = 0;
+static std::string device_id("AUTO");
```
agree, i'll  move it

sivanov-work commented on modules/gapi/src/streaming/onevpl/engine/transcode/transcode_engine_legacy.cpp:
```diff
@@ -0,0 +1,477 @@
+        return true;
+    }
+    return false;
+}
```
`Type` deducted from string `param_name` - so there would be concretization in caller code anyway like switch-case. 
By the way - i cannot see an approach how it may be written shorter ( and faster :))

OrestChura replied:
No, I meant
```
template<typename Type> Type get_vpp_param_value(MapIterator& it);

template<> uint16_t get_vpp_param_value<uint16_t>(MapIterator& it) {
    return it->second.Data.U16;
}

template<> uint32_t get_vpp_param_value<uint32_t>(MapIterator& it) {
    return it->second.Data.U32;
}

template<typename Type>
bool set_vpp_param(const char* name, Type& out_vpp_param,
                   const std::map<std::string, mfxVariant> &params_storage,
                   mfxSession session) {
    auto it = params_storage.find(name);
    if (it != params_storage.end()) {
        auto value = get_vpp_param_value<Type>(it);
        GAPI_LOG_INFO(nullptr, "[" << session << "] set \"" << name <<
                               "\": " << value);
        out_vpp_param = value;
        return true;
    }
    return false;
}
```

I believe <Type> can do its work, without switch-case

sivanov-work replied:
great, thanks

sivanov-work commented on modules/gapi/src/streaming/onevpl/engine/transcode/transcode_engine_legacy.cpp:
```diff
@@ -0,0 +1,477 @@
+            }
+            return ExecutionStatus::Continue;
+        },
+        // 4) transcode
```
it is, thanks

sivanov-work replied:
fixed

sivanov-work commented on modules/gapi/src/streaming/onevpl/engine/transcode/transcode_engine_legacy.cpp:
```diff
@@ -0,0 +1,477 @@
+    mfxVPPParams.vpp.Out = mfxVPPParams.vpp.In;
+
+    std::map<std::string, mfxVariant> cfg_vpp_params =
+                        VPLLegacyTranscodeEngine::get_vpp_params(cfg_params);
```
i prefer to emphasize  that it is static function by adding class name in prefix: help to read code

OrestChura replied:
OK, it does actually help, agree

sivanov-work commented on modules/gapi/src/streaming/onevpl/engine/transcode/transcode_session.cpp:
```diff
@@ -0,0 +1,70 @@
+        // Delegate exception processing on caller
+        throw;
+    }
+}
```
it process different surfaces & different pool_keys.
If you have idea how to unify it - please share it )

OrestChura replied:
So, in `LegacyDecodeSession` you can create a protected method:  
`void swap_surface(EngineSession& engine, VPLAccelerationPolicy::pool_key_t pool_id, std::weak_ptr<Surface> surface_ptr);`

which contains almost all the code:
```
void LegacyDecodeSession::swap_surface(ProcessingEngineBase& engine,
                                       VPLAccelerationPolicy::pool_key_t pool_id,
                                       std::weak_ptr<Surface> surface_ptr) {
    VPLAccelerationPolicy* acceleration_policy = engine.get_accel();
    GAPI_Assert(acceleration_policy && "Empty acceleration_policy");
    try {
        auto cand = acceleration_policy->get_free_surface(pool_id).lock();

        GAPI_LOG_DEBUG(nullptr, "[" << session << "] swap surface"
                                ", old: " << (!surface_ptr.expired()
                                              ? surface_ptr.lock()->get_handle()
                                              : nullptr) <<
                                ", new: "<< cand->get_handle());

        surface_ptr = cand;
    } catch (const std::runtime_error& ex) {
        GAPI_LOG_WARNING(nullptr, "[" << session << "] error: " << ex.what());

        // Delegate exception processing on caller
        throw;
    }
}
```

and then for different Sessions create their own public methods (to keep your API). For Decode:
```
void swap_processing_surface(VPLLegacyDecodeEngine& engine);

void LegacyDecodeSession::swap_processing_surface(VPLLegacyDecodeEngine& engine) {
    swap_surface(engine, // will it be cast right to base class reference or this should be cast explicitly?..
                 decoder_pool_id, procesing_surface_ptr);
}
``` 

For Transcode:
```
void swap_transcode_surface(VPLLegacyTranscodeEngine& engine);

void LegacyTranscodeSession::swap_transcode_surface(VPLLegacyTranscodeEngine& engine) {
    swap_surface(engine, // will it be cast right to base class reference or this should be cast explicitly?..
                 vpp_out_pool_id, vpp_surface_ptr);
}
``` 

There could be mistakes, sorry. But I believe the idea might work

sivanov-work replied:
I think it may work with little modifications, thanks

mpashchenkov commented on modules/gapi/src/streaming/onevpl/cfg_params.cpp:
```diff
@@ -118,6 +118,83 @@ CfgParam CfgParam::create_implementation(const char* value) {
+                            static_cast<uint64_t>(value), false);
+}
+
+
```
Extra empty line.

sivanov-work replied:
fixed

mpashchenkov reviewed (commented):
Some unanswered questions.

sivanov-work commented on modules/gapi/src/streaming/onevpl/accelerators/accel_policy_dx11.cpp:
```diff
@@ -261,24 +259,55 @@ mfxStatus VPLDX11AccelerationPolicy::on_alloc(const mfxFrameAllocRequest *reques
+    std::vector<ComPtrGuard<ID3D11Texture2D>> main_textures;
+    main_textures.reserve(main_textures_count);
+    for (size_t i = 0; i < main_textures_count; i++) {
+
```
fixed

sivanov-work commented on modules/gapi/src/streaming/onevpl/engine/decode/decode_engine_legacy.hpp:
```diff
@@ -24,20 +24,31 @@ struct DecoderParams;
+
+    SessionParam prepare_session_param(mfxSession mfx_session,
+                                        const std::vector<CfgParam>& cfg_params,
+                                        std::shared_ptr<IDataProvider> provider);
```
fixed

dmatveev commented on modules/gapi/samples/onevpl_infer_single_roi.cpp:
```diff
@@ -294,11 +323,25 @@ int main(int argc, char *argv[]) {
+                     " For exploring list of supported transformations please find out "
+                     " vpp_* related stuff in"
+                     " gapi/include/opencv2/gapi/streaming/onevpl/cfg_params.hpp\n"
+                     " Pay attention that to obtain expected result In this case VPP "
```
no `\n` here

dmatveev commented on modules/gapi/samples/onevpl_infer_single_roi.cpp:
```diff
@@ -294,11 +323,25 @@ int main(int argc, char *argv[]) {
+
+        std::cout <<"/*******************************************************/\n"
+                    "ATTENTION: GPU Inference Engine preprocessing is not vital as expected!\n"
+                     " Please consider param \"source_preproc_enable=1\" and specify "
+                     " appropriated media frame transformation using oneVPL::VPP primitives"
```
no `\n` here as well

dmatveev commented on modules/gapi/samples/onevpl_infer_single_roi.cpp:
```diff
@@ -294,11 +323,25 @@ int main(int argc, char *argv[]) {
+                     " Please consider param \"source_preproc_enable=1\" and specify "
+                     " appropriated media frame transformation using oneVPL::VPP primitives"
+                     " which force onevpl::GSource to produce tranformed media frames.\n"
+                     " For exploring list of supported transformations please find out "
+                     " vpp_* related stuff in"
```
wow here too

dmatveev commented on modules/gapi/src/backends/ie/giebackend.cpp:
```diff
@@ -510,15 +510,51 @@ inline IE::Blob::Ptr extractRemoteBlob(IECallContext& ctx, std::size_t i) {
 
-    ParamType* blob_params = cv::util::any_cast<ParamType>(&any_blob_params);
-    if (blob_params == nullptr) {
+    ParamType* frame_blob_params = cv::util::any_cast<ParamType>(&any_blob_params);
+    if (frame_blob_params == nullptr) {
         GAPI_Assert(false && "Incorrect type of blobParams: "
                               "expected std::pair<InferenceEngine::TensorDesc,"
                                                  "InferenceEngine::ParamMap>");
     }
```
so there should be `bad_any_cast` already so no need to use `*`

dmatveev reviewed (approved):
_Approved with comments_

OrestChura reviewed (approved):
Some remaining comments can be answered and addressed in the next PR

sivanov-work commented:
it seems before merge need to rebase with latter nv12 changes in giebackend

alalek commented:
> dmatveev added this to the **category: g-api** milestone 6 hours ago

WTF???
Again, No new milestones please.

sivanov-work commented on modules/gapi/src/backends/ie/giebackend.cpp:
```diff
@@ -510,15 +510,51 @@ inline IE::Blob::Ptr extractRemoteBlob(IECallContext& ctx, std::size_t i) {
-    return ctx.uu.rctx->CreateBlob(blob_params->first,
-                                   blob_params->second);
+    if (ctx.uu.params.device_id.find("GPU") != std::string::npos) {
+        GAPI_LOG_DEBUG(nullptr, "Extract remote blob for GPU DX11 device_id: " <<
```
no MediaFrame Adapter for it - DX11 only
There is a check with assert whilst we create Source that available device type

sivanov-work commented:
@alalek  Could you merge it please? all un-addressed comments would by applied in the next PR https://github.com/opencv/opencv/pull/21363/
with some little refactoring
